File: packages/sdk-node/src/transport.ts
import type { TraceIngestEnvelope } from '@audit-trail/schema';
import { HttpResponseError, isRetriableHttpStatus, withRetry } from './retry';

export interface BufferedTransportConfig {
  apiKey: string;
  projectId: string;
  baseUrl?: string;
  /** Flush at least every N ms (default 2000). */
  flushIntervalMs?: number;
  /** Flush when buffer reaches this size (default 50). */
  maxBufferSize?: number;
  fetchImpl?: typeof fetch;
}

export interface TraceSink {
  enqueue(envelope: TraceIngestEnvelope): void;
  flush(): Promise<void>;
  shutdown(): Promise<void>;
}

/**
 * Buffers trace envelopes and flushes to POST /v1/traces with retry + jitter.
 */
export class BufferedTransport implements TraceSink {
  private readonly config: Required<
    Pick<BufferedTransportConfig, 'baseUrl' | 'flushIntervalMs' | 'maxBufferSize'>
  > &
    BufferedTransportConfig;
  private readonly fetchImpl: typeof fetch;
  private readonly buffer: TraceIngestEnvelope[] = [];
  private timer: ReturnType<typeof setInterval> | null = null;
  private flushing = false;

  constructor(config: BufferedTransportConfig) {
    this.config = {
      ...config,
      baseUrl: config.baseUrl ?? 'http://localhost:3000',
      flushIntervalMs: config.flushIntervalMs ?? 2000,
      maxBufferSize: config.maxBufferSize ?? 50,
    };
    this.fetchImpl = config.fetchImpl ?? fetch;
    this.timer = setInterval(() => {
      void this.flush();
    }, this.config.flushIntervalMs);
  }

  enqueue(envelope: TraceIngestEnvelope): void {
    const sanitized = maybeRedact(envelope);
    const idx = this.buffer.findIndex((t) => t.trace_id === sanitized.trace_id);
    if (idx >= 0) {
      this.buffer[idx] = sanitized;
    } else {
      this.buffer.push(sanitized);
    }
    if (this.buffer.length >= this.config.maxBufferSize) {
      void this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.flushing || this.buffer.length === 0) {
      return;
    }
    this.flushing = true;
    const batch = this.buffer.splice(0, this.buffer.length);
    try {
      for (const envelope of batch) {
        await this.sendOne(envelope);
      }
    } finally {
      this.flushing = false;
    }
  }

  async shutdown(): Promise<void> {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush();
  }

  private async sendOne(envelope: TraceIngestEnvelope): Promise<void> {
    await withRetry(async () => {
      const response = await this.fetchImpl(`${this.config.baseUrl}/v1/traces`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Api-Key': this.config.apiKey,
          'X-Project-Id': this.config.projectId,
          'Idempotency-Key': envelope.trace_id,
        },
        body: JSON.stringify(envelope),
      });
      if (!response.ok && response.status !== 202) {
        const body = await response.text();
        throw new HttpResponseError(response.status, `Ingest failed (${response.status}): ${body}`);
      }
    });
  }
}

export function isRetriableStatus(status: number): boolean {
  return isRetriableHttpStatus(status);
}

function maybeRedact(envelope: TraceIngestEnvelope): TraceIngestEnvelope {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { redactSecrets } = require('./redact') as typeof import('./redact');
    return redactSecrets(envelope);
  } catch {
    return envelope;
  }
}

File: packages/sdk-python/audit_trail/transport.py
from __future__ import annotations

import json
import random
import threading
import time
import urllib.error
import urllib.request
from typing import Any, Callable, Optional

from .redact import redact_secrets

RETRIABLE_STATUS = {429, 500, 502, 503, 504}


class BufferedTransport:
    def __init__(
        self,
        *,
        api_key: str,
        project_id: str,
        base_url: str = "http://localhost:3000",
        flush_interval_ms: int = 2000,
        max_buffer_size: int = 50,
        post_fn: Optional[Callable[[dict[str, Any]], None]] = None,
    ) -> None:
        self.api_key = api_key
        self.project_id = project_id
        self.base_url = base_url.rstrip("/")
        self.max_buffer_size = max_buffer_size
        self._post_fn = post_fn or self._default_post
        self._buffer: list[dict[str, Any]] = []
        self._lock = threading.Lock()
        self._timer = threading.Timer(flush_interval_ms / 1000.0, self._tick)
        self._timer.daemon = True
        self._timer.start()

    def enqueue(self, envelope: dict[str, Any]) -> None:
        sanitized = redact_secrets(envelope)
        with self._lock:
            for idx, existing in enumerate(self._buffer):
                if existing["trace_id"] == sanitized["trace_id"]:
                    self._buffer[idx] = sanitized
                    break
            else:
                self._buffer.append(sanitized)
            should_flush = len(self._buffer) >= self.max_buffer_size
        if should_flush:
            self.flush()

    def flush(self) -> None:
        with self._lock:
            batch = self._buffer[:]
            self._buffer.clear()
        for envelope in batch:
            self._send_with_retry(envelope)

    def shutdown(self) -> None:
        self._timer.cancel()
        self.flush()

    def _tick(self) -> None:
        self.flush()

    def _send_with_retry(self, envelope: dict[str, Any], max_attempts: int = 5) -> None:
        delay = 0.25
        for attempt in range(1, max_attempts + 1):
            try:
                self._post_fn(envelope)
                return
            except RuntimeError as exc:
                if attempt >= max_attempts or not _is_retriable(str(exc)):
                    raise
                time.sleep(random.uniform(0, min(8.0, delay)))
                delay = min(8.0, delay * 2)

    def _default_post(self, envelope: dict[str, Any]) -> None:
        data = json.dumps(envelope).encode("utf-8")
        request = urllib.request.Request(
            f"{self.base_url}/v1/traces",
            data=data,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "X-Api-Key": self.api_key,
                "X-Project-Id": self.project_id,
                "Idempotency-Key": envelope["trace_id"],
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                if response.status not in (200, 202):
                    raise RuntimeError(f"Ingest failed ({response.status})")
        except urllib.error.HTTPError as exc:
            if exc.code in RETRIABLE_STATUS:
                raise RuntimeError(f"retriable {exc.code}") from exc
            raise RuntimeError(f"Ingest failed ({exc.code})") from exc


def _is_retriable(message: str) -> bool:
    return "retriable" in message or "429" in message or "503" in message
